src/python/pants/testutil/engine/util.py

```diff
@@ -143,6 +144,55 @@ def configurations(self):
 TARGET_TABLE = SymbolTable({'struct': Struct, 'target': Target})
 
 
+def rootify_rules(
+  *rules: Union[Callable, TaskRule, RootRule, UnionRule]
+) -> Tuple[Union[Callable, RootRule, UnionRule], ...]:
+  """Utility to convert `optionable_rule`s into `RootRule`s and no-op on other types of rules.
+
+  This is useful for `TestBase.request_single_product_request`-style tests during the setup of
+  `rules()`. To use an `optionable_rule` in tests, you have to provide an `OptionsBootstrapper`,
+  which is more heavy-weight than most tests want. This prevents certain tests from using the
```

Comment:
> which is more heavy-weight than most tests want

Is this true? I expect that most tests would be ok with options parsing occurring?

Manually injecting all of the subsystems deps isn't going to make sense in the average cases, I think... it feels like any time you used this utility (because you have 4 or more sets of options to inject) it would probably be a better idea to use options parsing...?

Reply:
Hm. This comment turns out to be misleading in its focus on `OptionsBootstrapper`. I started working on your proposal and found we don't even need to touch `OptionsBootstrapper` to get these 3 tests to pass without `rootify_rules`. This is the diff that it would take:

```diff
diff --git a/src/python/pants/backend/python/lint/black/rules_integration_test.py b/src/python/pants/backend/python/lint/black/rules_integration_test.py
index 5a805f264..066ac266f 100644
--- a/src/python/pants/backend/python/lint/black/rules_integration_test.py
+++ b/src/python/pants/backend/python/lint/black/rules_integration_test.py
@@ -24,7 +24,6 @@ from pants.engine.selectors import Params
 from pants.rules.core.fmt import FmtResult
 from pants.rules.core.lint import LintResult
 from pants.source.wrapped_globs import EagerFilesetWithSpec
-from pants.testutil.engine.util import rootify_rules
 from pants.testutil.subsystem.util import global_subsystem_instance, init_subsystems
 from pants.testutil.test_base import TestBase
 
@@ -37,15 +36,19 @@ class BlackIntegrationTest(TestBase):
 
   @classmethod
   def rules(cls):
-    return rootify_rules(
+    return (
       *super().rules(),
       *black_rules(),
       *pex_rules(),
       *download_pex_bin_rules(),
       *python_native_code_rules(),
       *subprocess_environment_rules(),
+      RootRule(Black),
       RootRule(BlackSetup),
       RootRule(FormattablePythonTarget),
+      RootRule(PythonNativeCode),
+      RootRule(PythonSetup),
+      RootRule(SubprocessEnvironment),
     )
 
   def setUp(self):
diff --git a/src/python/pants/backend/python/rules/pex_test.py b/src/python/pants/backend/python/rules/pex_test.py
index 5bd6da651..300b66cef 100644
--- a/src/python/pants/backend/python/rules/pex_test.py
+++ b/src/python/pants/backend/python/rules/pex_test.py
@@ -25,7 +25,6 @@ from pants.engine.fs import Digest, DirectoryToMaterialize, FileContent, InputFi
 from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
 from pants.engine.rules import RootRule
 from pants.engine.selectors import Params
-from pants.testutil.engine.util import rootify_rules
 from pants.testutil.subsystem.util import init_subsystems
 from pants.testutil.test_base import TestBase
 from pants.util.strutil import create_path_env_var
@@ -35,13 +34,16 @@ class TestResolveRequirements(TestBase):
 
   @classmethod
   def rules(cls):
-    return rootify_rules(
+    return (
       *super().rules(),
       *pex_rules(),
       *download_pex_bin_rules(),
       *python_native_code_rules(),
       *subprocess_environment_rules(),
       RootRule(CreatePex),
+      RootRule(PythonNativeCode),
+      RootRule(PythonSetup),
+      RootRule(SubprocessEnvironment),
     )
 
   def setUp(self):
diff --git a/src/python/pants/rules/core/strip_source_roots_test.py b/src/python/pants/rules/core/strip_source_roots_test.py
index 2cdbbb9e4..4a01629cb 100644
--- a/src/python/pants/rules/core/strip_source_roots_test.py
+++ b/src/python/pants/rules/core/strip_source_roots_test.py
@@ -11,7 +11,6 @@ from pants.engine.selectors import Params
 from pants.rules.core.strip_source_root import SourceRootStrippedSources
 from pants.rules.core.strip_source_root import rules as strip_source_root_rules
 from pants.source.source_root import SourceRootConfig
-from pants.testutil.engine.util import rootify_rules
 from pants.testutil.subsystem.util import init_subsystem
 from pants.testutil.test_base import TestBase
 
@@ -19,8 +18,12 @@ from pants.testutil.test_base import TestBase
 class StripSourceRootsTests(TestBase):
   @classmethod
   def rules(cls):
-    return rootify_rules(
-      *super().rules(), *create_fs_rules(), *strip_source_root_rules(), RootRule(HydratedTarget),
+    return (
+      *super().rules(),
+      *create_fs_rules(),
+      *strip_source_root_rules(),
+      RootRule(SourceRootConfig),
+      RootRule(HydratedTarget),
     )
 
   def setUp(self):
```

--

However, this is still very compelling evidence IMO for why we should still add `rootify_rules`. We can cut out all of that above diff by using `rootify_rules` to leverage the fact that the production code already is registering these subsystems. With `rootify_rule`, we avoid having to register any given subsystem both in production and in tests.

--

I'll update this comment and the PR description.

(Also, FYI the original reason I was talking about `OptionsBootstrapper`: https://pantsbuild.slack.com/archives/C0D7TNJHL/p1573784090071900. It seems that sometimes that is true that we would indeed need to initialize an `OptionsBootstrapper`.)